### ISSUE TITLE
CertificateChainValidationEngine certificate format

### ISSUE THREAD

**Issue by god830novo** [PeculiarVentures/PKI.js#228]
What format should I use for rootCertificate in
`const trustedCerts = [rootCertificate]`?
I have my certificates in base64 format, basically I got rid of ----BEGINCERTIFICATE----- and -----ENDCERTIFICATE-----
I tried to parse to certificate, base64and ArrayBuffer and all 3 formats got me this error.
Error:
{result: false, resultCode: -1, resultMessage: "Cannot read property 'byteLength' of undefined"}

**Comment by rmhrisk** [comment]
You can see an example of certificate validation here:
https://github.com/PeculiarVentures/PKI.js/blob/master/examples/CertificateComplexExample/es6.js#L479

You will see it is DER in that sample:
https://github.com/PeculiarVentures/PKI.js/blob/master/examples/CertificateComplexExample/es6.js#L544

**Comment by god830novo** [comment]
@rmhrisk Hi Ryan I believe I misunderstood the buffer part.
Can you show me the format of a buffer that works for the following code please?
I tried solution here https://github.com/PeculiarVentures/PKI.js/issues/58 but it parsed base64 into some reasonable words and some '<d~µw<»Q-'2À×A^»U úÞ.'
```
const asn1 = asn1js.fromBER(buffer1);
const certificate1 = new pkijs.Certificate({ schema: asn1.result });
```

**Comment by god830novo** [comment]
@rmhrisk Hi Ryan, is there a method that allow me to verify signature using certificate? I have both signature and certificate in base64. I couldn't find it in this repo. Thanks!

**Comment by rmhrisk** [comment]
It depends on what kind of signature.

If for example you have a raw RSA signature you want to first take the public key from the certificate (which you can do with PKIjs) and them import into webcrypto so you can validate the signature was created with that corresponding private key.

See: https://github.com/diafygi/webcrypto-examples

Then you would validate the certificate chain to see if you trust the subject it represents.

It is possible your signature is in a different format, for example a PKCS7 if so that has its own proceeding rules you need to be concerned with.

**Comment by god830novo** [comment]
@rmhrisk Ryan where is the example for taking the public key from the certificate?
I checked this one: https://github.com/PeculiarVentures/PKI.js/issues/158
but got more confused.

**Comment by rmhrisk** [comment]
Here is some sample code that reads the certificate public key and hashes it:

‘’’
const fingerprint = await crypto.digest("SHA-1", certificate.subjectPublicKeyInfo.toSchema().toBER(false));
‘’’

You can modify this to suit your use case.

**Comment by god830novo** [comment]
@rmhrisk Can you tell me what's wrong with this code please?
I'm able to get public key in arraybuffer(ArrayBuffer(32)) by your code you provided.
But it failed at importing, I can't import the arraybuffer to a CryptoKey format key.
I tried different format and input when importkey, but none of them works. 
<pre>
window.crypto.subtle.digest("SHA-256", signingCertificate.subjectPublicKeyInfo.toSchema().toBER(false))
  .then((res) => {
    const signingCertificatePublicKeyBuffer = res;
    window.crypto.subtle.importKey(
      "spki",
      signingCertificatePublicKeyBuffer,
      { 
        name: "RSASSA-PKCS1-v1_5",
        hash: {name: "SHA-256"},
      },
      true,
      ["verify"]
    )
    .then(publicKey => {
      console.log(publicKey);
    })
    .catch((err) => {
      console.error(err);
    });
  })
  .catch((err) => {
    console.error(err);
  });
</pre>

Thanks for you help!

**Comment by YuryStrozhevsky** [comment]
@god830novo As for this question:

> Hi Ryan, is there a method that allow me to verify signature using certificate?

So, if you need to verify signature inside the certificate you simply need this code:
```javascript
const yourBase64Certificate = "blahgblahstring";

const asn1 = asn1js.fromBER(stringToArrayBuffer(fromBase64(yourBase64Certificate)));

const certificate = new pkijs.Certificate({ schema: asn1.result });

const verifySignatureResult = await certificate.verify();
```
The `certificate.verify()` has a parameter `issuerCertificate` and call with empty value would work **only** for `self-signed` certificates. In case it is not a `self-signed` certificate you need to pass to `certificate.verify` the `issuer certificate` (it has the same `pkijs.Certificate` type).

So, if you need to verify any other signatures, for example inside CMS SignedData you need to call another methods, in another classes. For SignedData you would need [**this example**](https://github.com/PeculiarVentures/PKI.js/tree/master/examples/CMSSignedComplexExample). Or you could use [**live example**](https://pkijs.org/examples/CMSSigned_complex_example.html) with the same functionality.

In conclusion: the question `is there a method that allow me to verify signature using certificate?` is not so correct because in order to give you any correct answer I need to know where exactly would be the signature you would like to verify. It could be in certificate, SignedData, OCSP response, TSP response, whereever. But PKIjs has specialized functions for **any** of these cases, you do not need to write code with exporting public keys from certificate. I do recommend you to check [**main examples folder**](https://github.com/PeculiarVentures/PKI.js/tree/master/examples).

**Comment by god830novo** [comment]
@rmhrisk Nevermind, I made a mistake, shouldn't hash it!
